contrib/common.sh

```diff
@@ -0,0 +1,80 @@
+# Common functions for wrapper scripts for xdg-desktop-portal-termfilechooser.
+
+# Input args and ouptut format of the wrapper scripts:
+#
+# Inputs:
+# 1. "1" if multiple files can be chosen, "0" otherwise.
+# 2. "1" if a directory should be chosen, "0" otherwise.
+# 3. "0" if opening files was requested, "1" if writing to a file was
+#    requested. For example, when uploading files in Firefox, this will be "0".
+#    When saving a web page in Firefox, this will be "1".
+# 4. If writing to a file, this is recommended path provided by the caller. For
+#    example, when saving a web page in Firefox, this will be the recommended
+#    path Firefox provided, such as "~/Downloads/webpage_title.html".
+#    Note that if the path already exists, we keep appending "_" to it until we
+#    get a path that does not exist.
+# 5. The output path, to which results should be written.
+#
+# Output:
+# The script should print the selected paths to the output path (argument #5),
+# one path per line.
+# If nothing is printed, then the operation is assumed to have been canceled.
+
+
+default_termcmd () {
+	# Picking the default terminal emulator. Similar to
+	# 	https://github.com/i3/i3/blob/next/i3-sensible-terminal
+	# The first available command is used.
+	wtitle='FileChooser'  # Title text for the terminal window
+	term_cmds="
+		x-terminal-emulator -e
+		wezterm -e
+		kitty --title="$wtitle"
+		gnome-terminal --title="$wtitle" --wait --
+		urxvt -title "$wtitle" -e
+		uxterm -title "$wtitle" -e
+		xterm -title "$wtitle" -e
+		"
+	echo "$term_cmds" | while read -r tcmd; do
+		if command -v "${tcmd%% *}" > /dev/null 2>&1 ; then
+			echo "$tcmd"
+			break
+		fi
+	done
+}
+
+
+create_save_file() {
+	path=$1
+	out=$2
+	cat > "$path" <<-'END'
+	xdg-desktop-portal-termfilechooser saving files tutorial
+
+	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+	!!!                 === WARNING! ===                 !!!
+	!!! The contents of *whatever* file you open last in !!!
+	!!! file manager will be *overwritten*!              !!!
+	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+	Instructions:
+	1) Move this file wherever you want.
+	2) Rename the file if needed.
+	3) Confirm your selection by opening the file, for
+		 example by pressing <Enter>.
+
+	Notes:
+	1) This file is provided for your convenience. You
+		 could delete it and choose another file to overwrite
+		 that, for example.
+	2) If you quit file manager without opening a file, this
+	   file will be removed and the save operation aborted.
+	END
+	rm "$out" || : 	# Remove the old /tmp/termfilechooser.portal
```

Comment:
This one is ok, but not necessary since file from /tmp won't persist after machine restarted.

Reply:
Termfilechooser [does not remove its temp file](https://github.com/GermainZ/xdg-desktop-portal-termfilechooser/issues/14). This tempfile contains the path of the "selection". This would be a problem on the next run, if for some reason it is not overwritten (e.g. the terminal exits without giving a choice), then the old "selected" file location will be used.

Reply:
Yep, we already have a function to remove the old file in filchooser.c, but I will write an if statement to check if the portal file exists and have the write permission or not. If not, it will return and print a log.

Reply:
I have not examined the C code. The remove-tempfile function likely does not get called (as the mentioned upstream issue still occurs).

Reply:
Make sure to run `systemctl --user daemon-reload && systemctl restart --user xxx` for both the xdg-portal and term portal.
Then, create the `/tmp/termfilechooser.portal` file and change its ownership to the root user or another user, so you don't have read or write permissions.
Finally, run `xdg-desktop-portal-termfilechooser -l ERROR -r` to view the error log. You can now test with Zenity.

The portal file is removed before each run, so if you don't have read or write permissions to the file, it will crash immediately, even if the file containts old selected data.

Reply:
❯ echo '/initial/dummy/value' > /tmp/termfilechooser.portal
	❯ sudo chown root:root /tmp/termfilechooser.portal
	❯ GTK_USE_PORTAL=1  zenity --file-selection   # select any file
	/initial/dummy/value
	❯ cat /tmp/termfilechooser.portal
	/initial/dummy/value